tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs

```diff
@@ -0,0 +1,234 @@
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(key));
+            }
+
+            var tags = AsList();
+
+            lock (tags)
+            {
+                bool tagsModified = false;
+
+                for (int i = 0; i < tags.Count; i++)
+                {
+                    if (string.Equals(tags[i].Key, key, StringComparison.Ordinal))
+                    {
+                        if (value == null)
+                        {
+                            tags.RemoveAt(i);
+                            tagsModified = true;
+                        }
+                        else if (!string.Equals(tags[i].Value, value, StringComparison.Ordinal))
+                        {
+                            tags[i] = new KeyValuePair<string, string>(key, value);
+                            tagsModified = true;
+                        }
+
+                        // clear the cached header if we make any changes to a distributed tag
+                        if (tagsModified && key.StartsWith(PropagatedTagPrefix, StringComparison.Ordinal))
+                        {
+                            _cachedPropagationHeader = null;
+                        }
+
+                        return;
+                    }
+                }
+
+                // If we get there, the tag wasn't in the collection
+                if (value != null)
+                {
+                    tags.Add(new KeyValuePair<string, string>(key, value));
+
+                    // clear the cached header if we make any changes to a distributed tag
+                    if (key.StartsWith(PropagatedTagPrefix, StringComparison.Ordinal))
+                    {
+                        _cachedPropagationHeader = null;
+                    }
+                }
+            }
+        }
+
+        public string? GetTag(string key)
+        {
+            var tags = AsList();
+
+            lock (tags)
+            {
+                for (int i = 0; i < tags.Count; i++)
+                {
+                    if (string.Equals(tags[i].Key, key, StringComparison.Ordinal))
+                    {
+                        return tags[i].Value;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets a collection of propagated internal Datadog tags,
+        /// formatted as "key1=value1,key2=value2".
+        /// </summary>
+        public string ToPropagationHeader()
+        {
+            if (_cachedPropagationHeader == null)
+            {
+                // cache the propagated tags in a format ready
+                // for headers in case we need it multiple times
+                Interlocked.CompareExchange(ref _cachedPropagationHeader, FormatPropagationHeader(), null);
+            }
+
+            return _cachedPropagationHeader;
+        }
+
+        private string FormatPropagationHeader()
+        {
+            if (_tags.Count == 0)
```

Comment:
I think we should do the same to localVar + lock(tags)  kind of things in this method as well.

Reply:
Good catch on the missing lock, thanks. Added in bb427617e3b7b50901a36f8758ff66a86669481f.

Do you think we still need a local var?

Reply:
True, `_tags` is readonly, we don't need a local var to ensure we are working always with the same instance.